python/tskit/ms.py

```diff
@@ -0,0 +1,114 @@
+#
+# MIT License
+#
+# Copyright (c) 2020 Tskit Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Write the genotypes from tree sequences to ms-style output
+"""
+
+
+class msWriter:
+    """
+    Writes a ms representation of the genotypes in tree sequence to a
+    file-like object.
+    """
+
+    def __init__(
+        self,
+        tree_sequence,
+        print_trees=False,
+        precision=4,
+        mutation_rate=0,
+        num_replicates=1,
+        write_header=True,
+    ):
+        self._tree_sequence = tree_sequence
+        self._num_replicates = num_replicates
+        self._mutation_rate = mutation_rate
+        self._precision = precision
+        self._print_trees = print_trees
+        self._write_header = write_header
+
+    def print_trees(self, output):
+        """
+        Print out the trees in ms-format from the specified tree sequence.
+        When 'invisible' recombinations occur ms prints out copies of the
+        same tree. Therefore, we must keep track of all breakpoints from the
+        simulation and write out a tree for each one.
+        """
+        tree = next(self._tree_sequence.trees())
+        newick = tree.newick(precision=self._precision)
+        print(f"[{tree.span}]", newick, file=output)
+
+    def __write_header(self, output):
+        print(
+            "ms {} {}".format(
+                self._tree_sequence.sample_size, max(self._num_replicates, 1)
+            ),
```

Comment:
These two are the first two arguments in the ms file format. Including the rest is a complicated procedure, since the tree_sequence could be generated from different software like msprime or SLiM, and it is not straightforward to recreate all the ms-style command line arguments from that. I discussed this with @petrelharp a while ago and we thought that only having these two basic arguments makes sense.

Reply:
The question is: what do downstream users expect to see here? If no-one is parsing this line, then it would make sense to put something about provenance here, as @benjeffery suggests. But I do worry that people might be parsing it a bit (well, hopefully just looking for a line starting with `ms X Y` to skip), and they *might* be pulling the sample size and number of replicates out of it. So, I guess I vote for leaving it as-is?